<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 4.1.1">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css">


<script id="hexo-configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    hostname: new URL('http://yoursite.com').hostname,
    root: '/',
    scheme: 'Gemini',
    version: '7.6.0',
    exturl: false,
    sidebar: {"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},
    copycode: {"enable":true,"show_result":true,"style":null},
    back2top: {"enable":true,"sidebar":false,"scrollpercent":false},
    bookmark: {"enable":false,"color":"#222","save":"auto"},
    fancybox: false,
    mediumzoom: false,
    lazyload: false,
    pangu: false,
    comments: {"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},
    algolia: {
      appID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    },
    localsearch: {"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},
    path: 'search.xml',
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}
  };
</script>

  <meta name="description" content="SAM 是用来存储核苷酸序列比对信息的文件格式。SAM Tools 工具包提供各种工具处理SAM文件。包括功能：sorting, merging, indexing and generating alignments。安装教程见https:&#x2F;&#x2F;www.jianshu.com&#x2F;p&#x2F;53de170927a7 安装过程中有许多依赖的库需要安装的，可能每个人缺的库都不尽相同，不懂的就百度一下吧： 123">
<meta property="og:type" content="article">
<meta property="og:title" content="samtools 工具处理SAM文件">
<meta property="og:url" content="http://yoursite.com/2020/03/02/samtools/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="SAM 是用来存储核苷酸序列比对信息的文件格式。SAM Tools 工具包提供各种工具处理SAM文件。包括功能：sorting, merging, indexing and generating alignments。安装教程见https:&#x2F;&#x2F;www.jianshu.com&#x2F;p&#x2F;53de170927a7 安装过程中有许多依赖的库需要安装的，可能每个人缺的库都不尽相同，不懂的就百度一下吧： 123">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/14137386-2d3511483ee87a04?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/14137386-c09696c7b20a9f14?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">
<meta property="article:published_time" content="2020-03-02T08:06:58.000Z">
<meta property="article:modified_time" content="2020-03-02T08:06:58.100Z">
<meta property="article:author" content="Aimi">
<meta property="article:tag" content="生信软件">
<meta property="article:tag" content="序列比对">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://upload-images.jianshu.io/upload_images/14137386-2d3511483ee87a04?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240">

<link rel="canonical" href="http://yoursite.com/2020/03/02/samtools/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome: false,
    isPost: true
  };
</script>

  <title>samtools 工具处理SAM文件 | Hexo</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-meta">

    <div>
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Hexo</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
  </div>

  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>
</div>


<nav class="site-nav">
  
  <ul id="menu" class="menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-fw fa-home"></i>首页</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-fw fa-archive"></i>归档</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-fw fa-tags"></i>标签</a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-fw fa-th"></i>分类</a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>搜索
        </a>
      </li>
  </ul>

</nav>
  <div class="site-search">
    <div class="popup search-popup">
    <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocorrect="off" autocapitalize="none"
           placeholder="搜索..." spellcheck="false"
           type="text" id="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result"></div>

</div>
<div class="search-pop-overlay"></div>

  </div>
</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content">
            

  <div class="posts-expand">
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block " lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2020/03/02/samtools/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="Aimi">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Hexo">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          samtools 工具处理SAM文件
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2020-03-02 16:06:58" itemprop="dateCreated datePublished" datetime="2020-03-02T16:06:58+08:00">2020-03-02</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E5%B8%B8%E7%94%A8%E8%BD%AF%E4%BB%B6/" itemprop="url" rel="index">
                    <span itemprop="name">生物信息学常用软件</span>
                  </a>
                </span>
            </span>

          
            <span id="/2020/03/02/samtools/" class="post-meta-item leancloud_visitors" data-flag-title="samtools 工具处理SAM文件" title="阅读次数">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">阅读次数：</span>
              <span class="leancloud-visitors-count"></span>
            </span>
  
  <span class="post-meta-item">
    
      <span class="post-meta-item-icon">
        <i class="fa fa-comment-o"></i>
      </span>
      <span class="post-meta-item-text">Valine：</span>
    
    <a title="valine" href="/2020/03/02/samtools/#comments" itemprop="discussionUrl">
      <span class="post-comments-count valine-comment-count" data-xid="/2020/03/02/samtools/" itemprop="commentCount"></span>
    </a>
  </span>
  
  

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p><img src="https://upload-images.jianshu.io/upload_images/14137386-2d3511483ee87a04?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="image"></p>
<p>SAM 是用来存储核苷酸<code>序列比对信息</code>的文件格式。SAM Tools 工具包提供各种工具<code>处理SAM文件</code>。包括<code>功能</code>：sorting, merging, indexing and generating alignments。安装教程见<a href="https://www.jianshu.com/p/53de170927a7" target="_blank" rel="noopener">https://www.jianshu.com/p/53de170927a7</a></p>
<p>安装过程中有许多<code>依赖的库</code>需要安装的，可能每个人缺的库都不尽相同，不懂的就百度一下吧：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-get update</span><br><span class="line">sudo apt install gcc</span><br><span class="line">sudo apt install git</span><br><span class="line">sudo apt install make</span><br><span class="line">sudo apt-get install libncurses5-dev  ##bam_tview_curses.c:41:20: fatal error: curses.h: No such file or directory</span><br><span class="line">sudo yum install bzip2-devel  ##cram&#x2F;cram_io.c:57:19: fatal error: bzlib.h: No such file or directory</span><br><span class="line">sudo apt-get install liblzma-dev ##cram&#x2F;cram_io.c:60:18: fatal error: lzma.h: No such file or directory</span><br></pre></td></tr></table></figure>
<p>装好之后有如下这么多命令，下面我们只介绍samtools：</p>
<a id="more"></a>
<p><img src="https://upload-images.jianshu.io/upload_images/14137386-c09696c7b20a9f14?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="image"></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">rqq@ubuntu:~&#x2F;bio$ samtools</span><br><span class="line"></span><br><span class="line">Program: samtools (Tools for alignments in the SAM format)Version: 1.5 (using htslib 1.5) </span><br><span class="line">Usage:   samtools &lt;command&gt; [options] </span><br><span class="line">Commands:  -- Indexing</span><br><span class="line">     dict           create a sequence dictionary file</span><br><span class="line">     faidx          index&#x2F;extract FASTA</span><br><span class="line">     index          index alignment</span><br><span class="line">   -- Editing</span><br><span class="line">     calmd          recalculate MD&#x2F;NM tags and &#39;&#x3D;&#39; bases</span><br><span class="line">     fixmate        fix mate information</span><br><span class="line">     reheader       replace BAM header</span><br><span class="line">     rmdup          remove PCR duplicates</span><br><span class="line">     targetcut      cut fosmid regions (for fosmid pool only)</span><br><span class="line">     addreplacerg   adds or replaces RG tags</span><br><span class="line">   -- File operations</span><br><span class="line">     collate        shuffle and group alignments by name     cat            concatenate BAMs</span><br><span class="line">     merge          merge sorted alignments</span><br><span class="line">     mpileup        multi-way pileup     sort           sort alignment file</span><br><span class="line">     split          splits a file by read group</span><br><span class="line">     quickcheck     quickly check if SAM&#x2F;BAM&#x2F;CRAM file appears intact</span><br><span class="line">     fastq          converts a BAM to a FASTQ</span><br><span class="line">     fasta          converts a BAM to a FASTA</span><br><span class="line">   -- Statistics</span><br><span class="line">     bedcov         read depth per BED region</span><br><span class="line">     depth          compute the depth</span><br><span class="line">     flagstat       simple stats</span><br><span class="line">     idxstats       BAM index stats</span><br><span class="line">     phase          phase heterozygotes</span><br><span class="line">     stats          generate stats (former bamcheck)   -- Viewing</span><br><span class="line">     flags          explain BAM flags</span><br><span class="line">     tview          text alignment viewer</span><br><span class="line">     view           SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion</span><br><span class="line">     depad          convert padded BAM to unpadded BAM</span><br></pre></td></tr></table></figure>

<h2 id="一、samtools-view功能"><a href="#一、samtools-view功能" class="headerlink" title="一、samtools view功能"></a>一、samtools view功能</h2><h3 id="1、将sam文件转化为bam文件"><a href="#1、将sam文件转化为bam文件" class="headerlink" title="1、将sam文件转化为bam文件"></a>1、将sam文件转化为bam文件</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -bS in.sam &gt; in.bam</span><br></pre></td></tr></table></figure>
<p>是view的一个应用-b指定输出文件为bam, -S 指定输入文件为sam</p>
<h3 id="2、查看bam文件的header信息"><a href="#2、查看bam文件的header信息" class="headerlink" title="2、查看bam文件的header信息"></a>2、查看bam文件的header信息</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -H in.bam</span><br></pre></td></tr></table></figure>

<h3 id="3、取出bam文件的一部分"><a href="#3、取出bam文件的一部分" class="headerlink" title="3、取出bam文件的一部分"></a>3、取出bam文件的一部分</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -b your.bam Chr1 &gt; Chr1.bam</span><br></pre></td></tr></table></figure>

<ul>
<li>-b <strong>指定是输出文件</strong>为bam, </li>
<li>Chr1指定你<strong>要看的是哪一部分</strong>， 这里指看Chr1那一部分，然后重定向到一个新的bam文件，注意，这个bam文件是没有header的，如果想要包括header 可以使用-h参数。</li>
</ul>
<p><strong>随机取出bam文件的某一部分出来， 必须要有index文件</strong>，否则会报错： [bam_index_load] fail to load BAM index. [main_samview] random alignment retrieval only works for indexed BAM files.</p>
<p><strong>关于view更详细的参数说明</strong>：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">Usage:   samtools view [options] &lt;in.bam&gt;|&lt;in.sam&gt; [region1 [...]]</span><br><span class="line">Options:</span><br><span class="line">-b       output BAM         </span><br><span class="line">-h       print header for the SAM output         </span><br><span class="line">-H       print header only (no alignments)         </span><br><span class="line">-S       input is SAM         </span><br><span class="line">-u       uncompressed BAM output (force -b)         </span><br><span class="line">-1       fast compression (force -b)         </span><br><span class="line">-x       output FLAG in HEX (samtools-C specific)         </span><br><span class="line">-X       output FLAG in string (samtools-C specific)         </span><br><span class="line">-c       print only the count of matching records         </span><br><span class="line">-B       collapse the backward CIGAR operation         </span><br><span class="line">-@ INT   number of BAM compression threads [0]         </span><br><span class="line">-L FILE  output alignments overlapping the input BED FILE [null]         </span><br><span class="line">-t FILE  list of reference names and lengths (force -S) [null]         </span><br><span class="line">-T FILE  reference sequence file (force -S) [null]         </span><br><span class="line">-o FILE  output file name [stdout]         </span><br><span class="line">-R FILE  list of read groups to be outputted [null]         </span><br><span class="line">-f INT   required flag, 0 for unset [0]         </span><br><span class="line">-F INT   filtering flag, 0 for unset [0]         </span><br><span class="line">-q INT   minimum mapping quality [0]         </span><br><span class="line">-l STR   only output reads in library STR [null]         </span><br><span class="line">-r STR   only output reads in read group STR [null]         </span><br><span class="line">-s FLOAT fraction of templates to subsample; integer part as seed [-1]        </span><br><span class="line">  -?       longer help</span><br></pre></td></tr></table></figure>

<h3 id="4、bam文件-转化为sam文件"><a href="#4、bam文件-转化为sam文件" class="headerlink" title="4、bam文件 转化为sam文件"></a>4、bam文件 转化为sam文件</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -h R12.merged.bam &gt; test.sam</span><br></pre></td></tr></table></figure>
<p>-h是将bam文件中的<strong>header也加入到sam文件</strong>中。 比如htseq-count老版本只接受sam文件</p>
<h2 id="二、建立索引Indexing"><a href="#二、建立索引Indexing" class="headerlink" title="二、建立索引Indexing"></a>二、建立索引Indexing</h2><p>注意</p>
<ul>
<li><p>如果你执行命令的地方和<strong>参考序列不在同一个目录</strong>，参考序列用全路径，最后的<code>index结果和参考序列在同一个</code>目录里面，而不是执行命令的目录。</p>
</li>
<li><p>在fasta文件中，对于某一个序列，除了最后一行，其他行所含碱基数应该一样。</p>
</li>
</ul>
<h3 id="1、对fasta文件建立索引"><a href="#1、对fasta文件建立索引" class="headerlink" title="1、对fasta文件建立索引"></a>1、对fasta文件建立索引</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools faidx ref.fasta`</span><br></pre></td></tr></table></figure>

<h3 id="2、对BAM文件建立索引"><a href="#2、对BAM文件建立索引" class="headerlink" title="2、对BAM文件建立索引"></a>2、对BAM文件建立索引</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools index in.bam</span><br></pre></td></tr></table></figure>
<ul>
<li>结果文件名为in.bam.bai</li>
</ul>
<h3 id="3、知道位置信息查找对应的序列信息"><a href="#3、知道位置信息查找对应的序列信息" class="headerlink" title="3、知道位置信息查找对应的序列信息"></a>3、知道位置信息查找对应的序列信息</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools faidx ref.fa Chr1:33667-33667</span><br></pre></td></tr></table></figure>
<p>指查看染色体一上的第33667个碱基。</p>
<h2 id="三、将bam文件进行sort"><a href="#三、将bam文件进行sort" class="headerlink" title="三、将bam文件进行sort"></a>三、将bam文件进行sort</h2><p>只能对<strong>bam文件</strong>进行sort, 不能对sam文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools sort aln.bam anl.sorted</span><br></pre></td></tr></table></figure>
<ul>
<li>默认是根据<strong>coordinate</strong>进行sort， 如果输入bam文件为in.bam , 则输出文件名为in.sorted.bam </li>
<li>如果要按照<strong>read name</strong>进行sort, 需要加<code>-n</code>, 如heseq-count 就要求文件时按照read name 而不是coordinate。<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools sort -n aln.bam anl.sorted</span><br></pre></td></tr></table></figure>

</li>
</ul>
<h2 id="四、去除bam文件中pcr导致的重复reads信息"><a href="#四、去除bam文件中pcr导致的重复reads信息" class="headerlink" title="四、去除bam文件中pcr导致的重复reads信息"></a>四、去除bam文件中pcr导致的重复reads信息</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools rmdup in.bam in.rmp.bam</span><br></pre></td></tr></table></figure>

<h2 id="五、合并bam文件"><a href="#五、合并bam文件" class="headerlink" title="五、合并bam文件"></a>五、合并bam文件</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools merge out.bam in1.bam in2.bam in3.bam</span><br></pre></td></tr></table></figure>
<p> 假如in1.bam, in2.bam, in3.bam是某个某样本的三个重复，我们可以将他们合并为一个bam文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools merge -R chr1 out.bam in1.bam in2.bam in3.bam</span><br></pre></td></tr></table></figure>
<p> 如果想对<strong>部分合并</strong>，如至合并一号染色的上的bam文件合并，chr1必须为序列的名字，一号染色体序列的名字为Chr1，那么就应为<strong>-R Chr1</strong></p>
<p><strong>注意</strong>：要合并的bam文件，必须有对应的index文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools index in.bam  <span class="comment">#结果文件名为in.bam.bai</span></span><br></pre></td></tr></table></figure>




<h2 id="六、统计bam文件中的reads情况，有多少reads比对上"><a href="#六、统计bam文件中的reads情况，有多少reads比对上" class="headerlink" title="六、统计bam文件中的reads情况，有多少reads比对上"></a>六、统计bam文件中的reads情况，有多少reads比对上</h2><p>命令： </p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools flagstat RF12.merged.bam</span><br></pre></td></tr></table></figure>
<p>结果如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">66196754 + 0 in total (QC-passed reads + QC-failed reads)</span><br><span class="line">#bam文件中所有的reads数。</span><br><span class="line"></span><br><span class="line">0 + 0 duplicates</span><br><span class="line">66196754 + 0 mapped (100.00%:-nan%)</span><br><span class="line">#比对到基因组上的reads数目， 我用的比对软件是tophat， 结果中只保留比对上的reads信息。</span><br><span class="line"></span><br><span class="line">66196754 + 0 paired in sequencing</span><br><span class="line">#属于paired reads数目， 我的数据都是双端测序。所以都是paired reads。</span><br><span class="line"></span><br><span class="line">33493586 + 0 read1</span><br><span class="line">#来自于read1中的reads数目</span><br><span class="line"></span><br><span class="line">32703168 + 0 read2</span><br><span class="line">#来自于read2 中的reads数目</span><br><span class="line">#read1 + read2 &#x3D; paired reads</span><br><span class="line"></span><br><span class="line">45729393 + 0 properly paired (69.08%:-nan%)</span><br><span class="line">#完美匹配的reads数， 即一对reads比对到同一条参考序列，并且这对reads之间的举例符合设置的阈值</span><br><span class="line"></span><br><span class="line">61118410 + 0 with itself and mate mapped</span><br><span class="line">#一对reads 都比对到了参考序列上，但是不一定比对到同一条染色体</span><br><span class="line"></span><br><span class="line">5078344 + 0 singletons (7.67%:-nan%)</span><br><span class="line">#一对reads中，只有一条匹配到了参考序列上。</span><br><span class="line">#61118410+5078344&#x3D;66196754</span><br><span class="line"></span><br><span class="line">703397 + 0 with mate mapped to a different chr</span><br><span class="line">#一对reads比对到了不同的染色体上。针对的是with itself and mate mapped</span><br><span class="line"></span><br><span class="line">346693 + 0 with mate mapped to a different chr (mapQ&gt;&#x3D;5)</span><br><span class="line">#和上面一样，只不过比对的质量大于等于5的reads数目 &lt;&#x2F;pre&gt;</span><br></pre></td></tr></table></figure>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>本文作者： </strong>Aimi
  </li>
 
  <li class="post-copyright-license">
    <strong>版权声明： </strong>本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fa fa-fw fa-creative-commons"></i>BY-NC-SA</a> 许可协议。转载请注明出处！
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6/" rel="tag"><i class="fa fa-tag"></i> 生信软件</a>
              <a href="/tags/%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9/" rel="tag"><i class="fa fa-tag"></i> 序列比对</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2020/03/02/sam/" rel="prev" title="SAM文件格式介绍">
      <i class="fa fa-chevron-left"></i> SAM文件格式介绍
    </a></div>
      <div class="post-nav-item">
    <a href="/2020/03/02/hisat2/" rel="next" title="hisat2比对软件将reads比对到参考基因组">
      hisat2比对软件将reads比对到参考基因组 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  

  </div>


          </div>
          
    <div class="comments" id="valine-comments"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let activeClass = CONFIG.comments.activeClass;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#一、samtools-view功能"><span class="nav-number">1.</span> <span class="nav-text">一、samtools view功能</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1、将sam文件转化为bam文件"><span class="nav-number">1.1.</span> <span class="nav-text">1、将sam文件转化为bam文件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2、查看bam文件的header信息"><span class="nav-number">1.2.</span> <span class="nav-text">2、查看bam文件的header信息</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3、取出bam文件的一部分"><span class="nav-number">1.3.</span> <span class="nav-text">3、取出bam文件的一部分</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4、bam文件-转化为sam文件"><span class="nav-number">1.4.</span> <span class="nav-text">4、bam文件 转化为sam文件</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#二、建立索引Indexing"><span class="nav-number">2.</span> <span class="nav-text">二、建立索引Indexing</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1、对fasta文件建立索引"><span class="nav-number">2.1.</span> <span class="nav-text">1、对fasta文件建立索引</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2、对BAM文件建立索引"><span class="nav-number">2.2.</span> <span class="nav-text">2、对BAM文件建立索引</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3、知道位置信息查找对应的序列信息"><span class="nav-number">2.3.</span> <span class="nav-text">3、知道位置信息查找对应的序列信息</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#三、将bam文件进行sort"><span class="nav-number">3.</span> <span class="nav-text">三、将bam文件进行sort</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#四、去除bam文件中pcr导致的重复reads信息"><span class="nav-number">4.</span> <span class="nav-text">四、去除bam文件中pcr导致的重复reads信息</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#五、合并bam文件"><span class="nav-number">5.</span> <span class="nav-text">五、合并bam文件</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#六、统计bam文件中的reads情况，有多少reads比对上"><span class="nav-number">6.</span> <span class="nav-text">六、统计bam文件中的reads情况，有多少reads比对上</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">Aimi</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">20</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">15</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">23</span>
        <span class="site-state-item-name">标签</span></a>
      </div>
  </nav>
</div>
  <div class="cc-license motion-element" itemprop="license">
    <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" class="cc-opacity" rel="noopener" target="_blank"><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"></a>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2020</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Aimi</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> 强力驱动 v4.1.1
  </div>
  <span class="post-meta-divider">|</span>
  <div class="theme-info">主题 – <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a> v7.6.0
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  




  
<script src="/js/local-search.js"></script>













  

  


<script>
NexT.utils.loadComments(document.querySelector('#valine-comments'), () => {
  NexT.utils.getScript('//unpkg.com/valine/dist/Valine.min.js', () => {
    var GUEST = ['nick', 'mail', 'link'];
    var guest = 'nick,mail,link';
    guest = guest.split(',').filter(item => {
      return GUEST.includes(item);
    });
    new Valine({
      el: '#valine-comments',
      verify: true,
      notify: true,
      appId: 'BSMCdPIQASIJECRiSlflbXHG-gzGzoHsz',
      appKey: 'oDXoXQeCGTX5ancyoDnbkjm1',
      placeholder: "这里留言",
      avatar: 'mm',
      meta: guest,
      pageSize: '10' || 10,
      visitor: true,
      lang: '' || 'zh-cn',
      path: location.pathname,
      recordIP: false,
      serverURLs: ''
    });
  }, window.Valine);
});
</script>

</body>
</html>
